multiuser/keycloak/che-multiuser-keycloak-server/src/main/java/org/eclipse/che/multiuser/keycloak/server/AbstractKeycloakFilter.java

```diff
@@ -13,51 +13,119 @@
 
 import static org.eclipse.che.multiuser.machine.authentication.shared.Constants.MACHINE_TOKEN_KIND;
 
-import io.jsonwebtoken.ExpiredJwtException;
-import io.jsonwebtoken.Jwt;
+import com.auth0.jwk.GuavaCachedJwkProvider;
+import com.auth0.jwk.JwkException;
+import com.auth0.jwk.JwkProvider;
+import com.auth0.jwk.UrlJwkProvider;
+import com.google.common.annotations.VisibleForTesting;
+import io.jsonwebtoken.Claims;
+import io.jsonwebtoken.JwsHeader;
+import io.jsonwebtoken.JwtException;
+import io.jsonwebtoken.JwtParser;
 import io.jsonwebtoken.Jwts;
-import io.jsonwebtoken.MalformedJwtException;
-import io.jsonwebtoken.SignatureException;
+import io.jsonwebtoken.SigningKeyResolverAdapter;
+import java.net.MalformedURLException;
+import java.net.URL;
+import java.security.Key;
 import java.security.PublicKey;
-import javax.inject.Inject;
 import javax.servlet.Filter;
 import javax.servlet.FilterConfig;
-import javax.servlet.ServletException;
 import javax.servlet.http.HttpServletRequest;
-import org.eclipse.che.multiuser.machine.authentication.server.signature.SignatureKeyManager;
+import org.eclipse.che.multiuser.keycloak.shared.KeycloakConstants;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
 
 /**
  * Base abstract class for the Keycloak-related servlet filters.
  *
- * <p>In particular it defines commnon use-cases when the authentication / multi-user logic should
- * be skipped
+ * <p>In particular it defines common use-cases when the authentication / multi-user logic should be
+ * skipped
  */
 public abstract class AbstractKeycloakFilter implements Filter {
 
-  @Inject protected SignatureKeyManager signatureKeyManager;
+  protected final JwtParser jwtParser;
+  protected KeycloakSettings keycloakSettings;
+  protected JwkProvider jwkProvider;
+
+  private static final Logger LOG = LoggerFactory.getLogger(AbstractKeycloakFilter.class);
+
+  public AbstractKeycloakFilter(KeycloakSettings keycloakSettings, long allowedClockSkewSec)
+      throws MalformedURLException {
+    this.keycloakSettings = keycloakSettings;
+    String jwksUrl = keycloakSettings.get().get(KeycloakConstants.JWKS_ENDPOINT_SETTING);
+    if (jwksUrl != null) {
+      this.jwkProvider = new GuavaCachedJwkProvider(new UrlJwkProvider(new URL(jwksUrl)));
+    }
+    jwtParser =
```

Comment:
This field initialization looks quite heavy. I think it worth to inject it from a container. With a provider for example.